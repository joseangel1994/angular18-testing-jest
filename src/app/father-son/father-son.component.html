<h5>Son</h5>
<code id="client-code">{{ client | json}}</code>
<br>

@if (client) {
  <button id="btn-change" (click)="onChange(5)">Cambiar ID</button>
  <button id="btn-delete" (click)="onDelete()">Borrar cliente</button>
}
