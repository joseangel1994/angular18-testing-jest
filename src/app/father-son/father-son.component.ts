import {Component, EventEmitter, Input, Output} from '@angular/core';
import {JsonPipe} from "@angular/common";
import {Client} from "../interfaces";

@Component({
  selector: 'app-father-son',
  standalone: true,
  imports: [
    JsonPipe
  ],
  templateUrl: './father-son.component.html'
})
export class FatherSonComponent {

  @Input() client?: Client
  @Output() onDeleteClient = new EventEmitter()
  @Output() onClientUpdated = new EventEmitter<Client>()

  onDelete() {
    this.client = undefined
    this.onDeleteClient.emit()
  }

  onChange(newId: number) {
    if(!this.client) return

    this.client = {
      ...this.client,
      id: newId
    }

    this.onClientUpdated.emit(this.client)
  }

}
