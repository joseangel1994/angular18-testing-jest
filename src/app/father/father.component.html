<h2>Father</h2>
<button (click)="onSetClient('test')">Cargar cliente</button>
<br>

<code>{{ client | json}}</code>
<hr>

<app-father-son [client]="client" (onDeleteClient)="client = undefined" (onClientUpdated)="client = $event"></app-father-son>
