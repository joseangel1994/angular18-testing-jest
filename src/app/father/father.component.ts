import { Component } from '@angular/core';
import {JsonPipe} from "@angular/common";
import {FatherSonComponent} from "../father-son/father-son.component";
import {Client} from "../interfaces";

@Component({
  selector: 'app-father',
  standalone: true,
  imports: [
    JsonPipe,
    FatherSonComponent
  ],
  templateUrl: './father.component.html'
})
export class FatherComponent {
  public client?: Client

  onSetClient(name: string) {
    this.client = {id: 1, name}
  }

}
