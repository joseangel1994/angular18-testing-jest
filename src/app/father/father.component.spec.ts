import {ComponentFixture, TestBed} from '@angular/core/testing';

import {FatherComponent} from './father.component';
import {FatherSonComponent} from "../father-son/father-son.component";
import {By} from "@angular/platform-browser";

describe('FatherComponent', () => {
  let component: FatherComponent;
  let fixture: ComponentFixture<FatherComponent>;
  let compiled: HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [FatherComponent, FatherSonComponent]
    })
      .compileComponents();

    fixture = TestBed.createComponent(FatherComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot()
  })

  it('should set the client with the correct name', () => {
    component.onSetClient('Jose')
    fixture.detectChanges()

    const code = compiled.querySelector('#client-code')
    expect(code?.textContent).toContain('"name"')
    expect(code?.textContent).toContain('Jose')
  })

  it('should delete the client if onDeleteClient emitted', () => {
    component.client = {id: 1, name: 'Jose'}
    fixture.detectChanges()

    const sonDebugElement = fixture.debugElement.query(By.directive(FatherSonComponent))
    const sonComponent: FatherSonComponent = sonDebugElement.componentInstance

    sonComponent.onDeleteClient.emit()
    expect(component.client).toBe(undefined)
  })

  it('should change the client if onClientUpdated emitted', () => {
    component.client = {id: 1, name: 'Jose'}
    fixture.detectChanges()

    const sonDebugElement = fixture.debugElement.query(By.directive(FatherSonComponent))
    const sonComponent: FatherSonComponent = sonDebugElement.componentInstance

    sonComponent.onClientUpdated.emit({id: 10, name: 'Jose'})
    expect(component.client).toEqual({id: 10, name: 'Jose'})
  })
});
