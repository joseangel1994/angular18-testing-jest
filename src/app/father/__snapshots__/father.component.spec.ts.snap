// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <h2>
    Father
  </h2>
  <button>
    Cargar cliente
  </button>
  <br />
  <code>
    
  </code>
  <hr />
  <app-father-son>
    <h5>
      Son
    </h5>
    <code>
      
    </code>
    <br />
    
  </app-father-son>
</div>
`;
